src/string/is_valid_email_address.rs

```diff
@@ -0,0 +1,139 @@
+const MAX_DOMAIN_LENGTH: u32 = 255;
+const CHARACTERS: &str = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz1234567890";
+const SYMBOLS: &str = "!#$%&'*+-/=?^_`{|}~";
+
+fn clear_quotes(input: &str) -> String {
+    let mut quote_count = input.starts_with('"') as u32;
+    let mut new_local_part: String = String::from(input.chars().next().unwrap());
+    for i in 1..input.len() {
+        if input.chars().nth(i).unwrap() == '"' && input.chars().nth(i - 1).unwrap() != '\\' {
+            quote_count += 1;
+
+            if !new_local_part.starts_with('"') && quote_count != 1 {
+                new_local_part.push('"');
+            }
+        }
+
+        if quote_count % 2 == 0 {
+            new_local_part.push(input.chars().nth(i).unwrap());
+        }
+    }
+
+    new_local_part
+}
+
+/// Follows email address rules as listed:
+/// https://en.wikipedia.org/wiki/Email_address#Examples
+pub fn is_valid_email_address(input_email_address: &str) -> bool {
+    let email_address = clear_quotes(input_email_address);
+
+    let parts: Vec<String> = email_address
+        .split('@')
+        .map(|x| x.to_string())
+        .collect::<Vec<String>>();
+
+    // (1)  ensure there is only one '@' symbol in the address
+    if parts.len() != 2 {
+        return false;
+    }
+
+    let (local_part, domain): (String, String) = (parts[0].clone(), parts[1].clone());
+
+    // (2)  check that the length of the two parts are within the valid range
+    if local_part.len() > MAX_LOCAL_PART_LENGTH as usize
+        || domain.len() > MAX_DOMAIN_LENGTH as usize
+    {
+        return false;
+    }
+
+    // (3)  check quote dots
+    for i in 1..local_part.len() {
+        if local_part.chars().nth(i - 1).unwrap() == '"'
+            && local_part.chars().nth(i).unwrap() == '"'
+        {
+            let proceeding_quote = local_part.chars().nth(i + 1);
+            if proceeding_quote.is_some() && proceeding_quote.unwrap() != '.' {
+                return false;
+            }
+        }
+    }
+
+    // (4)  validate the dots in the local part
+    if local_part.starts_with('.') || local_part.ends_with('.') || local_part.contains("..") {
+        return false;
+    }
+
+    // (5)  check that the characters in the local part are valid
+    for item in local_part.chars() {
+        if !CHARACTERS.contains(item) && !SYMBOLS.contains(item) && item != '.' && item != '"' {
+            return false;
+        }
+    }
+
+    // (6)  check that the domain contains valid characters
+    for item in domain.chars() {
+        if !CHARACTERS.contains(item) && item != '-' && item != '.' && !"[]:".contains(item) {
+            return false;
+        }
+    }
+
+    // (7)  validate the dots in the domain
+    if domain.starts_with('.') || domain.ends_with('.') || domain.contains("..") {
+        return false;
+    }
+
+    // (8)  validate the dashes in the domain
+    if domain.starts_with('-') || domain.ends_with('-') {
+        return false;
+    }
+
+    true
```

Comment:
What do you think about introducing functions like
- `is_local_part_valid`, which would then split into `is_quoted_local_part_valid` and `is_unquoted_local_part_valid`,
- `is_domain_valid`, which would then split accordingly.

Reply:
Due to the way the code is written, checking the quoted part and unquoted part separately wouldn't make the code more readable, so instead I just refactored the code and split each piece of logic into smaller sections.